plugin/pkg/admission/resourcequota/controller.go

```diff
@@ -0,0 +1,490 @@
+		work:       map[string][]*admissionWaiter{},
+		dirtyWork:  map[string][]*admissionWaiter{},
+		inProgress: sets.String{},
+	}, nil
+}
+
+// Run begins watching and syncing.
+func (e *quotaEvaluator) Run(workers int) {
+	defer utilruntime.HandleCrash()
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(e.doWork, time.Second, make(chan struct{}))
+	}
+}
+
+func (e *quotaEvaluator) doWork() {
+	for {
+		func() {
+			ns, admissionAttributes := e.getWork()
+			defer e.completeWork(ns)
+			if len(admissionAttributes) == 0 {
+				return
+			}
+
+			e.checkAttributes(ns, admissionAttributes)
+		}()
+	}
+}
+
+func (e *quotaEvaluator) checkAttributes(ns string, admissionAttributes []*admissionWaiter) {
+	// notify all on exit
+	defer func() {
+		for _, admissionAttribute := range admissionAttributes {
+			close(admissionAttribute.finished)
+		}
+	}()
+
+	quotas, err := e.getQuotas(ns)
+	if err != nil {
+		for _, admissionAttribute := range admissionAttributes {
+			admissionAttribute.result = err
+		}
+		return
+	}
+	if len(quotas) == 0 {
+		for _, admissionAttribute := range admissionAttributes {
+			admissionAttribute.result = nil
+		}
+		return
+	}
+
+	e.checkQuotas(quotas, admissionAttributes, 3)
+}
+
+func (e *quotaEvaluator) checkQuotas(quotas []api.ResourceQuota, admissionAttributes []*admissionWaiter, remainingRetries int) {
+
+	// yet another copy to compare against originals to see if we actually have deltas
+	originalQuotas := make([]api.ResourceQuota, len(quotas), len(quotas))
+	copy(originalQuotas, quotas)
+
+	atLeastOneChanged := false
+	for i := range admissionAttributes {
+		admissionAttribute := admissionAttributes[i]
+		newQuotas, err := e.checkRequest(quotas, admissionAttribute.attributes)
+		if err != nil {
+			admissionAttribute.result = err
+			continue
+		}
+
+		// if the new quotas are the same as the old quotas, then this particular one doesn't issue any updates
+		// that means that no quota docs applied, so it can get a pass
+		atLeastOneChangeForThisWaiter := false
+		for j := range newQuotas {
+			if !reflect.DeepEqual(originalQuotas[j], newQuotas[j]) {
+				atLeastOneChanged = true
+				atLeastOneChangeForThisWaiter = true
+				break
+			}
+		}
+
+		if !atLeastOneChangeForThisWaiter {
+			admissionAttribute.result = nil
+		}
+		quotas = newQuotas
+	}
+
+	// if none of the requests changed anything, there's no reason to issue an update, just fail them all now
+	if !atLeastOneChanged {
+		return
+	}
+
+	// now go through and try to issue updates.  Things get a little weird here:
+	// 1. check to see if the quota changed.  If not, skip.
+	// 2. if the quota changed and the update passes, be happy
+	// 3. if the quota changed and the update fails, add the original to a retry list
+	var updatedFailedQuotas []api.ResourceQuota
+	var lastErr error
+	nonConflictError := false
+	for i := range quotas {
+		newQuota := quotas[i]
+		// if this quota didn't change, skip it
+		if reflect.DeepEqual(originalQuotas[i], newQuota) {
```

Comment:
is deep equal better or worse than just checking if that status.used is not equal using:
https://github.com/kubernetes/kubernetes/blob/master/pkg/quota/resources.go#L26